src/js_parser.zig

```diff
@@ -23102,677 +23059,355 @@ fn NewParser_(
-                        ) catch unreachable;
-                        // Normally, we'd have to grow that inner function's stmts list by one
-                        // But we can avoid that by just making them all use this same array.
-                        var curr_stmts = _stmts;
-
-                        // in debug: crash in the printer due to undefined memory
-                        // in release: print ";" instead.
-                        // this should never happen regardless, but i'm just being cautious here.
-                        if (comptime !Environment.isDebug) {
-                            @memset(_stmts, Stmt.empty());
-                        }
-
-                        // Second pass: move any imports from the part's stmts array to the new stmts
-                        var imports_list = curr_stmts[0..imports_count];
-                        curr_stmts = curr_stmts[imports_list.len..];
-                        var toplevel_stmts = curr_stmts[0..toplevel_stmts_count];
-                        curr_stmts = curr_stmts[toplevel_stmts.len..];
-                        var exports_from = curr_stmts[0..exports_from_count];
-                        curr_stmts = curr_stmts[exports_from.len..];
-                        // This is used for onSetExports
-                        var update_function_stmts = curr_stmts[0..named_exports_count];
-                        curr_stmts = curr_stmts[update_function_stmts.len..];
-                        var export_all_function_body_stmts = curr_stmts[0..named_exports_count];
-                        curr_stmts = curr_stmts[export_all_function_body_stmts.len..];
-                        // This is the original part statements + 1
-                        var part_stmts = curr_stmts;
-                        if (comptime Environment.allow_assert) assert(part_stmts.len == end_iife_stmts_count);
-                        var part_stmts_i: usize = 0;
-
-                        var import_list_i: usize = 0;
-                        var export_list_i: usize = 0;
-
-                        // We must always copy it into the new stmts array
-                        for (part.stmts) |stmt| {
-                            switch (stmt.data) {
-                                .s_import => {
-                                    imports_list[import_list_i] = stmt;
-                                    import_list_i += 1;
-                                },
-                                .s_export_star, .s_export_from => {
-                                    exports_from[export_list_i] = stmt;
-                                    export_list_i += 1;
-                                },
-                                else => {
-                                    part_stmts[part_stmts_i] = stmt;
-                                    part_stmts_i += 1;
-                                },
+                parts = try hmr_transform_ctx.finalize(p, parts);
```

Comment:
does `parts_end` need to be updated here, or is it supposed to be end of the original parts?

Reply:
oh, nvm, misread